File: tvrename/src/tvrename/subtitles/ReferenceSubtitleDownloader.scala
package tvrename.subtitles

import java.io.File

import cats.data.OptionT
import cats.effect._
import cats.implicits._
import com.github.dnbn.submerge.api.parser.SRTParser
import org.http4s._
import org.http4s.client.Client
import org.http4s.client.dsl.io._
import tvrename._
import tvrename.config._

import scala.util.Try
import scala.util.matching.Regex

trait ReferenceSubtitleDownloader {
  def download(): IO[Unit]
}

case class ValidSubtitleFile(fileName: String)

class ReferenceSubtitleDownloaderImpl(
  config: TVRenameConfig,
  jobConfig: RemuxJobConfig,
  openSubtitles: OpenSubtitles,
  httpClient: Client[IO],
  fileSystem: FileSystem,
  logger: Logger
) extends ReferenceSubtitleDownloader {

  private val targetFolder: String =
    s"${config.cacheFolder}/reference/${jobConfig.seriesName}/Season ${"%02d".format(jobConfig.seasonNumber.value)}"

  private val episodeCountFile: String = s"${targetFolder}/.episode-count"

  private val parser: SRTParser = new SRTParser

  private def expectedEpisodeCount: IO[Option[Int]] =
    for {
      exists <- fileSystem.exists(episodeCountFile)
      lines <- if (exists) fileSystem.readLines(episodeCountFile) else IO.pure(List.empty)
    } yield lines.headOption.map(_.toInt)

  private def actualEpisodeCount: IO[Int] =
    fileSystem.walk(targetFolder, recursive = false).map(_.count(f => f.endsWith(".srt")))

  override def download(): IO[Unit] =
    for {
      _ <- fileSystem.makeDirs(targetFolder)
      expected <- expectedEpisodeCount
      actual <- actualEpisodeCount
      _ <- downloadIfNeeded(expected, actual)
    } yield ()

  private def downloadIfNeeded(expected: Option[Int], actual: Int): IO[Unit] = {
    expected match {
      case Some(a) if a == actual => IO.unit
      case _ =>
        for {
          searchResults <- openSubtitles.search(jobConfig.seriesId, jobConfig.seasonNumber)
          // TODO: handle no episodes found for series/season
          lastEpisode = searchResults.foldLeft(0)((acc, result) => math.max(acc, result.episodeNumber.value))
          _ <- logger.debug(
            s"${jobConfig.seriesName} Season ${jobConfig.seasonNumber.value.toString} has ${lastEpisode.toString} episodes"
          )
          _ <- fileSystem.writeToFile(episodeCountFile, lastEpisode.toString)
          _ <- downloadAllEpisodes(searchResults)
        } yield ()
    }
  }

  private def downloadAllEpisodes(seasonSearchResults: List[EpisodeSearchResults]): IO[Unit] = {
    seasonSearchResults.traverse_ { episode =>
      val sortedSearchResults = episode.searchResults
        .sortBy(_.isWebDL)(Ordering[Boolean].reverse)
        .sortBy(_.Score)(Ordering[Double].reverse)

      val template = jobConfig.template
        .replace("[series]", jobConfig.seriesName)
        .replace("[season]", f"${jobConfig.seasonNumber.value}%02d")
        .replace("[episode]", f"${episode.episodeNumber.value}%02d")

      val targetFile = s"$targetFolder/$template.srt"

      val downloadAll: OptionT[IO, Unit] = for {
        exists <- OptionT.liftF(fileSystem.exists(targetFile))
        subtitleFile <- OptionT(if (!exists) downloadValidSubtitleForEpisode(sortedSearchResults) else IO.pure(None))
        _ <- OptionT.liftF(logger.debug(s"\t => $template.srt"))
        _ <- OptionT.liftF(fileSystem.rename(subtitleFile.fileName, targetFile))
      } yield ()

      downloadAll.value
    }
  }

  def downloadValidSubtitleForEpisode(searchResults: List[SearchResult]): IO[Option[ValidSubtitleFile]] = {
    searchResults.foldLeft[IO[Option[ValidSubtitleFile]]](IO.pure(None)) { (acc, next) =>
      acc.flatMap {
        case Some(value) => IO.pure(Some(value))
        case None        => downloadAndParseSubtitle(next)
      }
    }
  }

  def downloadAndParseSubtitle(searchResult: SearchResult): IO[Option[ValidSubtitleFile]] = {
    val tempFile = fileSystem.getTempFileName()
    for {
      _ <- logger.debug(searchResult.SubDownloadLink)
      uri <- Uri.fromString(searchResult.SubDownloadLink).liftTo[IO]
      request <- Method.GET(uri)
      _ <- httpClient.stream(request).flatMap(s => fileSystem.gunzipToFile(s.body, tempFile)).compile.drain
      attempt <- cleanupAndParse(tempFile)
    } yield attempt
  }

  private def cleanupAndParse(fileName: String): IO[Option[ValidSubtitleFile]] = {
    for {
      lines <- fileSystem.readLines(fileName)
      _ <- fileSystem.writeToFile(fileName, cleanupLines(lines))
      attempt <- Try { parser.parse(new File(fileName)) }.attempt.liftTo[IO]
      result <- if (attempt.isLeft) IO.pure(None) else IO.pure(Some(ValidSubtitleFile(fileName)))
    } yield result
  }

  private def cleanupLines(lines: Seq[String]): String = {
    // many subrip files are formatted with timestamps like 00:00:0,000 or 00:00:00,00 which isn't valid for the parser
    // we don't really care about timestamps, though, so we'll try to make it valid
    val badSeconds =
      new Regex("([\\d]{2}):([\\d]{2}):(\\d),([\\d]{2,3})", "hours", "minutes", "seconds", "milliseconds")
    val badMilliseconds =
      new Regex("([\\d]{2}):([\\d]{2}):([\\d]{2}),([\\d]{2})(?!\\d)", "hours", "minutes", "seconds", "milliseconds")
    lines
      .map(line =>
        badSeconds.replaceAllIn(
          line,
          m => s"${m group "hours"}:${m group "minutes"}:0${m group "seconds"},${m group "milliseconds"}"
        )
      )
      .map(line =>
        badMilliseconds.replaceAllIn(
          line,
          m => s"${m group "hours"}:${m group "minutes"}:${m group "seconds"},${m group "milliseconds"}0"
        )
      )
      .mkString("\n")
  }
}

File: Subtitles/ReferenceSubtitleDownloader.cs
using System.IO.Compression;
using System.Text;
using System.Text.RegularExpressions;
using SubtitlesParser.Classes;
using SubtitlesParser.Classes.Parsers;
using TvRename.Models;

namespace TvRename.Subtitles;

public partial class ReferenceSubtitleDownloader
{
    private readonly ILogger<ReferenceSubtitleDownloader> _logger;

    private readonly OpenSubtitlesApiClient _openSubtitlesApiClient;

    public ReferenceSubtitleDownloader(
        OpenSubtitlesApiClient openSubtitlesApiClient,
        ILogger<ReferenceSubtitleDownloader> logger)
    {
        _openSubtitlesApiClient = openSubtitlesApiClient;
        _logger = logger;
    }

    public async Task<int> Download(ValidatedParameters parameters, CancellationToken cancellationToken)
    {
        string referenceFolder = Path.Combine(parameters.Folder, ".tvrename", "reference");
        if (!Directory.Exists(referenceFolder))
        {
            Directory.CreateDirectory(referenceFolder);
        }

        Option<int> maybeExpectedEpisodeCount = await GetExpectedEpisodeCount(referenceFolder);
        foreach (int expectedEpisodeCount in maybeExpectedEpisodeCount)
        {
            return expectedEpisodeCount;
        }

        // search open subtitles
        Either<Exception, List<EpisodeSearchResults>> maybeResults =
            await _openSubtitlesApiClient.Search(parameters.Imdb, parameters.Season, cancellationToken);

        foreach (List<EpisodeSearchResults> results in maybeResults.RightToSeq())
        {
            if (results.Count == 0)
            {
                _logger.LogError(
                    "OpenSubtitles returned no results for show {ShowTitle} imdb {Imdb} season {SeasonNumber}",
                    parameters.Title,
                    parameters.Imdb,
                    parameters.Season);

                return 0;
            }

            int episodeCount = results.Max(e => e.EpisodeNumber) - results.Min(e => e.EpisodeNumber) + 1;
            _logger.LogInformation(
                "{ShowTitle} season {SeasonNumber} has {Count} episodes",
                parameters.Title,
                parameters.Season,
                episodeCount);
            await WriteEpisodeCount(referenceFolder, episodeCount);

            await DownloadEpisodes(referenceFolder, parameters.Title, parameters.Season, results, cancellationToken);

            return episodeCount;
        }

        return 0;
    }

    private static async Task<Option<int>> GetExpectedEpisodeCount(string referenceFolder)
    {
        string episodeCountFile = Path.Combine(referenceFolder, ".episode-count");
        if (File.Exists(episodeCountFile) && int.TryParse(
                await File.ReadAllTextAsync(episodeCountFile),
                out int episodeCount))
        {
            return episodeCount;
        }

        return None;
    }

    private static async Task WriteEpisodeCount(string referenceFolder, int episodeCount)
    {
        string episodeCountFile = Path.Combine(referenceFolder, ".episode-count");
        await File.WriteAllTextAsync(episodeCountFile, episodeCount.ToString());
    }

    private async Task DownloadEpisodes(
        string referenceFolder,
        string showTitle,
        int seasonNumber,
        List<EpisodeSearchResults> results,
        CancellationToken cancellationToken)
    {
        var parser = new SrtParser();

        using var client = new HttpClient();

        foreach (EpisodeSearchResults episode in results)
        {
            if (cancellationToken.IsCancellationRequested)
            {
                return;
            }

            IOrderedEnumerable<SearchResult> sorted = episode.SearchResults
                .OrderBy(s => s.IsWebDl ? 0 : 1)
                .ThenByDescending(s => s.Score);

            string targetFile = Path.Combine(
                referenceFolder,
                $"{showTitle} - s{seasonNumber:00}e{episode.EpisodeNumber:00}.srt");

            if (!File.Exists(targetFile))
            {
                foreach (SearchResult searchResult in sorted.Filter(s => !string.IsNullOrWhiteSpace(s.SubDownloadLink)))
                {
                    if (cancellationToken.IsCancellationRequested)
                    {
                        return;
                    }

                    _logger.LogInformation("Downloading {Url}", searchResult.SubDownloadLink);

                    string tempFile = Path.GetTempFileName();
                    Stream stream = await client.GetStreamAsync(
                        new Uri(searchResult.SubDownloadLink!),
                        cancellationToken);
                    await using (var fs = new FileStream(tempFile, FileMode.OpenOrCreate))
                    {
                        await using (var gzs = new GZipStream(stream, CompressionMode.Decompress))
                        {
                            await gzs.CopyToAsync(fs, cancellationToken);
                        }
                    }

                    string[] lines = await File.ReadAllLinesAsync(tempFile, cancellationToken);
                    string cleaned = CleanupLines(lines);
                    await File.WriteAllTextAsync(tempFile, cleaned, cancellationToken);

                    try
                    {
                        await using (var ms = new MemoryStream(Encoding.UTF8.GetBytes(cleaned)))
                        {
                            Option<List<SubtitleItem>> _ = parser.ParseStream(ms, Encoding.UTF8);
                        }
                    }
                    catch (Exception ex)
                    {
                        _logger.LogError(ex, "Failed to validate subtitle {TempFile}", tempFile);
                        continue;
                    }

                    _logger.LogInformation("Validated as {SubtitleFile}", Path.GetFileName(targetFile));
                    File.Move(tempFile, targetFile);
                    break;
                }
            }
        }
    }

    private string CleanupLines(IEnumerable<string> lines)
    {
        var sb = new StringBuilder();

        // many subrip files are formatted with timestamps like 00:00:0,000 or 00:00:00,00 which isn't valid for the parser
        // we don't really care about timestamps, though, so we'll try to make it valid
        foreach (string line in lines)
        {
            string nextLine = line;

            Match m1 = BadSeconds().Match(line);
            if (m1.Success)
            {
                nextLine = line.Replace(
                    m1.Value,
                    $"{m1.Groups[1].Value}:{m1.Groups[2].Value}:0{m1.Groups[3].Value},{m1.Groups[4].Value}");
            }

            Match m2 = BadMilliseconds().Match(line);
            if (m2.Success)
            {
                nextLine = line.Replace(
                    m2.Value,
                    $"{m2.Groups[1].Value}:{m2.Groups[2].Value}:{m2.Groups[3].Value},{m2.Groups[4].Value}0");
            }

            sb.AppendLine(nextLine);
        }

        sb.AppendLine();
        return sb.ToString();
    }

    [GeneratedRegex(@"([\d]{2}):([\d]{2}):([\d]{2}),([\d]{2})(?!\d)")]
    private static partial Regex BadMilliseconds();

    [GeneratedRegex(@"([\d]{2}):([\d]{2}):(\d),([\d]{2,3})")]
    private static partial Regex BadSeconds();
}
